enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,424 @@
+This is, because by default, no other slice is configured to have
+`io.latency` set.
+
+###### Critical workload protection
+
+Even critical pod workloads are run in burstable QoS class pods, thus
+at **General Availability** time they will be eligible to swap.
+However, swapping can lead to increased latencies and response times.
+For example, if a critical pod is depending on `LivenessProbe`s, then
+these checks can start to fail, once the pod is starting to swap.
+
+This is undesirable and can put a node or a cluster (i.e. if a critical
+Operator is affected) at risk.
+
+In order to prevent this problem, swap will be selectively disabled
+for pod using the two well-known [critical `priorityClass`es]:
+
+* `system-cluster-critical`
+* `system-node-critical`
+
+###### Node memory pressure handling
+
+Dealing with memory pressure on a node is differentiating the TP fom GA.
+
+* **Technology Preview** - `memory.high` is set on the `kubepods.slice`
+  in order to force the node to swap, once the memory is filling up.
+  Only once swap is full, the system will cross `memory.high` and trigger
+  soft evictions.
+
+  * Pro
+    * Simple to achieve.
+  * Con
+    * A lot of memory pressure has ot be present in order to trigger
+      soft eviction.
+
+* **General Availability** - Memory based soft and hard eviction is going to
+  be disabled, in favor of enabling swap based hard evictions, based on new
+  swap traffic and swap utilization eviction metrics.
+
+  * Pro
+    * Simpel mental model. With memory only, memory eviction is used.
+      With swap, swap eviction is used.
+    * [LLN] applies, because all pods share the nodes memory
+  * Con
+    * If there are no burstable QoS pods on a node, then no swapping
+      can take place, and no swap related signal will be triggered.
+      Only way to remove pressure is cgroup level OOM.
+      This is considered to be an edge case and highly unlikely.
+      Prometheus alerts for this edge case will be added.
+
+##### Differences between Technology Preview vs GA
+
+|                              | TP            | GP                 |
+|------------------------------|---------------|--------------------|
+| SWAP Provisioning            | MachineConfig | MachineConfig      |
+| SWAP Eligibility             | VM pods       | burstable QoS pods |
+| Node service protection      | Yes           | Yes                |
+| I/O saturation protection    | Yes           | Yes                |
+| Critical workload protection | No            | Yes                |
+| Memory pressure handling     | Memory based  | Swap based         |
+
+
+### Risks and Mitigations
+
+#### Phase 1
+
+| Risk                                       | Mitigation             |
+|--------------------------------------------|------------------------|
+| Miss details and introduce instability     | Limit to VM pods       |
+| Difficult transition from WASP to K8s SWAP | Explicit API to choose |
```

Comment:
Is this outdated? I thought we're not going to introduce any swap related APIs

Reply:
Dropped it for now.
We should revisit this topic.